### PR TITLE
man: Document that sd_journal_wait() and related calls only generate …

### PR DESCRIPTION
**Pull request by syyhao1994**
…events at most every 250ms
https://github.com/systemd/systemd/issues/17574

### CODE REVIEW

**Comment by vcaputo** [comment]
I don't think it makes sense to mention this systemd-journald::journal-file.c implementation detail here as if it's part of the public sd-journal API.

This is not a real-time system; there's already an implicit expectation of arbitrary, potentially lengthy delays separating wakeup-triggering events from the processes blocking on them running.  There's no need to start describing underlying sources of latency at every event monitoring API.

We're free to change that interval or eliminate the delay altogether in future versions.  And if your program somehow relies on this interval, it's almost certainly broken.  Consider that if one sd_journal_wait wakeup is delayed because UNIX, the subsequent wakeup may occur in less than 250ms if in the mean time systemd-journald produced more logs.  The interval is of no programmatic value to sd-journal callers, and to mention it anywhere here implies this is of some importance the developer should care about, which would be some form of misleading IMNSHO.

**Comment by keszybz** [comment]
+1 to what @vcaputo said. I don't think this needs to be documented. @syyhao1994 was there some particular issue you were trying to address here?

**Comment by NickYang24** [comment]
@vcaputo @keszybz 
https://github.com/systemd/systemd/issues/17574
The issue listed above has also confused me for a long time. As rsyslog use the abi for getting log, i think may be a document about the journald is not supposed to give RT guarantees is needed as poettering mentioned in the issue. And also it is not accurate for the original document that "It will synchronously wait until the journal gets changed" in the sd_journal_get_fd.xml.
If you think there is no needed to do so, just close the PR.
Thank you for your attention.

**Comment by syyhao1994** [comment]
@vcaputo @keszybz 
https://github.com/systemd/systemd/issues/17574
The issue listed above has also confused me for a long time. As rsyslog use the abi for getting log, i think may be a document about the journald is not supposed to give RT guarantees is needed as poettering mentioned in the issue. And also it is not accurate for the original document that "It will synchronously wait until the journal gets changed" in the sd_journal_get_fd.xml.
If you think there is no needed to do so, just close the PR.
Thank you for your attention.

**Comment by vcaputo** [comment]
I just don't think *this* is the place to mention it.  Maybe there could be a blurb in systemd-journald about change notifications being buffered.  The point of objecting to *this* PR is it's not an explicit part of the API, and there's already an implicit expectation of non-deterministic latency because the operating system isn't real-time.  It's not journald that's not real-time, it's unix; many of the IPC mechanisms coalesce events (multiple pipe writes get can be buffered then consumed by a single read for example) and the scheduler makes no latency guarantees.  It's just not a journald specific issue.

My argument above is basically that the sd-journal API shouldn't be tasked with informing developers that unix isn't a realtime system in general.  But having a .25s notification delay is putting us in that position more than a shorter delay would.

Why don't we just shorten the delay to something less human-perceivable?  .25s may be a bit too long to get lost in the noise, we could reduce it to e.g. .1s without significant overhead and that might help prevent such confusion.  We should just find the happy middle ground in the implementation before resorting to writing more docs papering over implementation details.

**Comment by poettering** [comment]
I'd be fine if we'd add a comment that journal change notifications are async, and that due to wake-up optimization additional latencies are introduced. I don't think we should be more precise than that however, i.e. not mention the 250ms, since that is an implementation detail, and we should have the ability to change that. for example for network backed file systems it might make sense to increase this substantially.

i.e. i think it's worth mentioning that there are additional latencies involved, but stop at that, don't reveal internals, don't make something documented behaviour that we want to keep freedom of control for.

**Comment by poettering** [comment]
@syyhao1994 if you rework the patch as I suggested, I think that#d be a compromise everyone could get behind?

**Comment by syyhao1994** [comment]
> 
> 
> @syyhao1994 if you rework the patch as I suggested, I think that#d be a compromise everyone could get behind?

I think it is reasonable to rework the patch as you said.
So how about the change now? Review please.
@poettering

File: man/sd_journal_get_fd.xml [hunk 1]
```diff
@@ -140,9 +140,10 @@ else {
...
     <function>sd_journal_get_timeout()</function> and
     <function>sd_journal_process()</function> is
-    <function>sd_journal_wait()</function>. It will synchronously wait
-    until the journal gets changed. The maximum time this call sleeps
-    may be controlled with the <parameter>timeout_usec</parameter>
+    <function>sd_journal_wait()</function>. It will asynchronously
+    wait until the journal gets changed as the wake-up optimization
+    for notifications are introduced with latency. The maximum time
+    this call sleeps may be controlled with the <parameter>timeout_usec</parameter>
     parameter. Pass <constant>(uint64_t) -1</constant> to wait
```
Comment (poettering) [inline man/sd_journal_get_fd.xml:147]: the latency is not added by sd_journal_wait(). it's already applied to sd_journal_get_events()/sd_journal_get_timeout().

Maybe leave the existing sentences in place but add an extra sentence to the end of the paragraph.

"Note that all of the above change notification interfaces will typically not notice change notifications instantly, but show some additional (though typically small) notification latency for multiple reasons: certain storage backends require time-based polling, in other cases wake-ups are optimized by coalescing events, in addition to OS-level IO/CPU scheduling latencies."

Or something like that. Could you rework this?

**Comment by poettering** [comment]
OK, to speed things up I prepped the PR I suggested now myself in #18633. Hope that settles it. Let's continue discussion there. Hope this works for you @syyhao1994?

**Comment by syyhao1994** [comment]
> OK, to speed things up I prepped the PR I suggested now myself in #18633. Hope that settles it. Let's continue discussion there. Hope this works for you @syyhao1994?

I think it make sense, thanks a lot.

**Pull request closed without merging**